File: 백준/Silver/24460. 특별상이라도 받고 싶어/특별상이라도 받고 싶어.cc
#include<bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef pair < ll, ll>pairll;

const ll maxN = 1300;


ll arr[maxN][maxN];
ll N;

ll solution(ll y, ll x, ll len) {
	if (len == 1) {
		return arr[y][x];
	}

	//두번째꺼 찾기
	if (len == 2) {
		vector<ll>vec = { arr[y][x],arr[y + 1][x],arr[y][x + 1],arr[y + 1][x + 1] };
		sort(vec.begin(), vec.end());
		return vec[1];
	}

	vector<ll> vec = { solution(y, x, len / 2), solution(y + len/2 , x, len / 2) , solution(y, x + len/2, len / 2) ,solution(y + len / 2, x + len / 2, len / 2) };
	sort(vec.begin(), vec.end());
	return vec[1];
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); 
	cout.tie(0);

	cin >> N;

	for (ll y = 0; y < N; y++) {
		for (ll x = 0; x < N; x++) {
			cin >> arr[y][x];
		}
	}
	cout << solution(0, 0, N);
	
	return 0;
}

File: 백준/Silver/24460. 특별상이라도 받고 싶어/특별상이라도 받고 싶어.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.Scanner;

public class Main {

	static final int maxSize = 1030;

	static int[][] arr = new int[maxSize][maxSize];
	static int N;
	static int dx[] = { 0, 1, 1, 0 };
	static int dy[] = { 0, 0, 1, 1 };

	static int solution(int y, int x, int len) {

		if (len == 1)
			return arr[y][x];

		ArrayList<Integer> vec = new ArrayList<>();

		for (int k = 0; k < 4; k++) {
			int cy = y + len * dy[k] / 2;
			int cx = x + len * dx[k] / 2;
			vec.add(solution(cy, cx, len / 2));
		}

		Collections.sort(vec);

		return vec.get(1);
	}

	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);

		N = scanner.nextInt();

		for (int y = 0; y < N; y++) {
			for (int x = 0; x < N; x++) {
				arr[y][x] = scanner.nextInt();
			}
		}

		System.out.println(solution(0, 0, N));

		scanner.close();
	}
}